be/src/runtime/routine_load/routine_load_task_executor.cpp

```diff
@@ -0,0 +1,174 @@
+        LOG(INFO) << "routine load task " << task.id << " has already been submitted";
+        return Status::OK;
+    }
+
+    // create the context
+    StreamLoadContext* ctx = new StreamLoadContext(_exec_env);
+    ctx->load_type = TLoadType::ROUTINE_LOAD;
+    ctx->load_src_type = task.type;
+    ctx->job_id = task.job_id;
+    ctx->id = UniqueId(task.id);
+    ctx->txn_id = task.txn_id;
+    ctx->db = task.db;
+    ctx->table = task.tbl;
+    ctx->label = task.label;
+    ctx->auth.auth_code = task.auth_code;
+
+    // the routine load task'txn has alreay began in FE.
+    // so it need to rollback if encounter error.
+    ctx->need_rollback = true;
+
+    // set source related params
+    switch (task.type) {
+        case TLoadSourceType::KAFKA:
+            ctx->kafka_info = new KafkaLoadInfo(task.kafka_load_info);
+            break;
+        default:
+            LOG(WARNING) << "unknown load source type: " << task.type;
+            delete ctx;
+            return Status("unknown load source type");
+    }
+
+    VLOG(1) << "receive a new routine load task: " << ctx->brief();
+    // register the task
+    ctx->ref();
+    _task_map[ctx->id] = ctx;
+    
+    // offer the task to thread pool
+    if (!_thread_pool->offer(
+            boost::bind<void>(&RoutineLoadTaskExecutor::exec_task, this, ctx,
+            [this] (StreamLoadContext* ctx) {
+                std::unique_lock<std::mutex> l(_lock);
+                _task_map.erase(ctx->id);
+                LOG(INFO) << "finished routine load task " << ctx->brief();
+                if (ctx->unref()) {
+                    delete ctx;
+                }
+            }))) {
+
+        // failed to submit task, clear and return
+        LOG(WARNING) << "failed to submit routine load task: " << ctx->brief();
+        _task_map.erase(ctx->id);
+        if (ctx->unref()) {
+            delete ctx;
+        }
+        return Status("failed to submit routine load task");
+
+    } else {
+        LOG(INFO) << "submit a new routine load task: " << ctx->brief()
+                  << ", current tasks num: " << _task_map.size();
+        return Status::OK;
+    }
+}
+
+void RoutineLoadTaskExecutor::exec_task(
+        StreamLoadContext* ctx, ExecFinishCallback cb) {
+
+    // create pipe and consumer
+    std::shared_ptr<StreamLoadPipe> pipe;
+    std::shared_ptr<DataConsumer> consumer;
+    switch (ctx->load_src_type) {
+        case TLoadSourceType::KAFKA:
+            pipe = std::make_shared<KafkaConsumerPipe>();
+            consumer = std::make_shared<KafkaDataConsumer>(
+                    ctx, std::static_pointer_cast<KafkaConsumerPipe>(pipe));
+            ctx->body_sink = pipe;
+            break;
+        default:
+            std::stringstream ss;
+            ss << "unknown routine load task type: " << ctx->load_type;
+            err_handler(ctx, Status::CANCELLED, ss.str());
+            cb(ctx);
+            return;
+    }
+
+#define HANDLE_ERROR(stmt, err_msg) \
```

Comment:
don't use macro here, you can create another function do following work, and this function call that function, and checks its return and handler error